File: common/src/main/java/net/impleri/playerskills/integration/kubejs/events/PlayerSkillChangedEventJS.java
package net.impleri.playerskills.integration.kubejs.events;

import dev.latvian.mods.kubejs.server.ServerEventJS;
import dev.latvian.mods.kubejs.util.UtilsJS;
import net.impleri.playerskills.api.Skill;
import net.impleri.playerskills.server.events.SkillChangedEvent;
import net.minecraft.world.entity.player.Player;

public class PlayerSkillChangedEventJS<T> extends ServerEventJS {
    private final SkillChangedEvent<T> event;

    public PlayerSkillChangedEventJS(SkillChangedEvent<T> event) {
        super(UtilsJS.staticServer);
        this.event = event;
    }

    public boolean getIsImproved() {
        var type = event.getType();

        if (type == null) {
            return false;
        }

        return type.getNextValue(event.getPrevious()) == event.getSkill().getValue();
    }

    public boolean getIsDegraded() {
        var type = event.getType();

        if (type == null) {
            return false;
        }

        return type.getPrevValue(event.getPrevious()) == event.getSkill().getValue();
    }

    public Skill<T> getSkill() {
        return event.getSkill();
    }

    public Skill<T> getPrevious() {
        return event.getPrevious();
    }

    public Player getPlayer() {
        return event.getPlayer();
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/integrations/kubejs/skills/PlayerSkillChangedEventJS.kt
package net.impleri.playerskills.integrations.kubejs.skills

import dev.latvian.mods.kubejs.server.ServerEventJS
import dev.latvian.mods.kubejs.util.UtilsJS
import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.events.SkillChangedEvent
import net.minecraft.world.entity.player.Player

class PlayerSkillChangedEventJS<T>(private val event: SkillChangedEvent<T>) : ServerEventJS(UtilsJS.staticServer) {
  fun getIsImproved(): Boolean {
    val type = event.type ?: return false
    return type.getNextValue(event.previous) == event.skill.value
  }

  fun getIsDegraded(): Boolean {
    val type = event.type ?: return false
    return type.getPrevValue(event.previous) == event.skill.value
  }

  fun getSkill(): Skill<T> {
    return event.skill
  }

  fun getPrevious(): Skill<T> {
    return event.previous
  }

  fun getPlayer(): Player {
    return event.player
  }
}
